### PR TITLE
Add at/numStride/indexStride

### PR DESCRIPTION
**Pull request by jmh530**
Facilitates easier access to slice contents, especially when not contiguous.

### CODE REVIEW

File: source/mir/ndslice/slice.d [hunk 1]
```diff
@@ -2544,6 +2544,51 @@ public:
...
+    {
+        assert(n < elementCount, "valStride: n must be less than elementCount");
+
+        static if (kind == Contiguous || (kind == Canonical && N == 1)) {
+            return n;
+        } else static if (kind == Universal && N == 1) {
+            return n * _strides[0];
+        } else {
+            return indexStride(numIndex(n));
+        }
```
Comment (9il) [inline source/mir/ndslice/slice.d:2584]: I am not sure this compiles, but the code is almost the same that was used in FlattenedIterator.
```d
    ptrdiff_t valStride()(ptrdiff_t n) @safe scope const
    {
        ptrdiff_t _shift;
        foreach_reverse (i; Iota!(1, N))
        {
            immutable v = n / ptrdiff_t(_lengths[i]);
            n %= ptrdiff_t(_lengths[i]);
            static if (i == S)
                _shift += n;
            else
                _shift += n * _strides[i];
            n = v;
        }
        _shift += n * _strides[0];
        return _shift;
    }
```

Comment (9il) [reply source/mir/ndslice/slice.d:2584]: btw, any reason to make it public?

Comment (jmh530) [reply source/mir/ndslice/slice.d:2584]: @9il Not really. `indexStride` was public and I tried to take the same approach. I can make changes private.

Comment (9il) [reply source/mir/ndslice/slice.d:2584]: It is ok to make it public, but I am not sure about naming, and we need docs.

File: source/mir/ndslice/slice.d [hunk 2]
```diff
@@ -2544,6 +2544,51 @@ public:
...
+    }
+
+    ///
+    auto ref at(size_t n) scope return @trusted
```
Comment (9il) [inline source/mir/ndslice/slice.d:2588]: Also needs more docs and more verbose name

Comment (jmh530) [reply source/mir/ndslice/slice.d:2601]: Happy to add more docs later today. Any suggestion on name? Maybe `access` or `accessFlat`? I am flexible.

Along the same lines, should gmean/hmean be renamed geometricMean/harmonicMean? Might need deprecation given that there have been releases...

File: source/mir/ndslice/slice.d (hunk 1 above)
Comment (jmh530) [reply source/mir/ndslice/slice.d:2597]: Same as below. Happy to add more docs later today. Flexible on naming.

Comment (9il) [reply source/mir/ndslice/slice.d:2597]: BTW, this code is for Canonical and Universal slices. Contiguous should just return the `n`

Comment (jmh530) [reply source/mir/ndslice/slice.d:2595]: Thanks.

File: source/mir/ndslice/slice.d (hunk 2 above)
Comment (9il) [reply source/mir/ndslice/slice.d:2599]: accessFlag looks good. Existing names in stat are good as well.

File: source/mir/ndslice/slice.d [hunk 3]
```diff
@@ -2545,6 +2545,62 @@ public:
         assert(mut == x);
     }
 
+    size_t[N] numIndex(size_t n) @safe scope const
```
Comment (9il) [inline source/mir/ndslice/slice.d:2548]: unused, can be removed though

**Pull request merged**